File: erl/keyword_ser.erl
-module(keyword_ser).
-behaviour(gen_server).

%% API
-export([start_link/1, add_keyword_position/3, do_query/1, get_ids/1]).

%% gen_server callbacks
-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
	 terminate/2, code_change/3]).

-define(SERVER, ?MODULE).

-record(state, {document_positions}).

%%%===================================================================
%%% API
%%%===================================================================

start_link(ServerName) ->
    gen_server:start_link({local, ServerName}, ?MODULE, [], []).
    

add_keyword_position(Keyword, DocCatId, Position) ->
    Server = get_keyword_server( Keyword ),
    gen_server:cast(Server, {add, DocCatId, Position}).

do_query(Terms) when is_list(Terms) ->
    DocumentMatches = lists:flatten(rpc:pmap( { keyword_ser, get_ids }, [], Terms)),
    MatchesPerTermAndDocument = lists:foldl( 
				  fun( {Term, DocIds}, AccIn1) ->
					  lists:foldl(fun(DocId, AccIn2) -> dict:update_counter( {Term, DocId}, 1, AccIn2) end, AccIn1, DocIds)
				  end, dict:new(), DocumentMatches),
    async_bm25:document_scores( Terms, DocumentMatches, MatchesPerTermAndDocument).

get_ids(Term)->
    case find_keyword_server( Term ) of
	{ ok, Server } ->
	    { Term, gen_server:call( Server, get_ids )};
	{ error, _ } ->
	    { Term, []}
    end.

%%%===================================================================
%%% gen_server callbacks
%%%===================================================================

%%--------------------------------------------------------------------
%% @spec init(Args) -> {ok, State} |
%%                     {ok, State, Timeout} |
%%                     ignore |
%%                     {stop, Reason}
%%--------------------------------------------------------------------
init([]) ->
    {ok, #state{document_positions=dict:new()}}.

%%--------------------------------------------------------------------
%% @spec handle_call(Request, From, State) ->
%%                                   {reply, Reply, State} |
%%                                   {reply, Reply, State, Timeout} |
%%                                   {noreply, State} |
%%                                   {noreply, State, Timeout} |
%%                                   {stop, Reason, Reply, State} |
%%                                   {stop, Reason, State}
%%--------------------------------------------------------------------
handle_call(get_ids, _From, State) ->
    Reply = dict:fold(
	      fun( DocId, Positions, AccIn) -> 
		      lists:duplicate(length( Positions ), DocId) ++ AccIn 
	      end,
	      [], State#state.document_positions
	     ),
    {reply, Reply, State}.

%%--------------------------------------------------------------------
%% @spec handle_cast(Msg, State) -> {noreply, State} |
%%                                  {noreply, State, Timeout} |
%%                                  {stop, Reason, State}
%%--------------------------------------------------------------------
handle_cast({add, DocCatId, Position}, State ) ->
    NewState = #state{ document_positions = dict:append( DocCatId, Position, State#state.document_positions ) },
    {noreply, NewState}.

%%--------------------------------------------------------------------
%% @spec handle_info(Info, State) -> {noreply, State} |
%%                                   {noreply, State, Timeout} |
%%                                   {stop, Reason, State}
%%--------------------------------------------------------------------
handle_info(_Info, State) ->
    {noreply, State}.

%%--------------------------------------------------------------------
%% @spec terminate(Reason, State) -> void()
%%--------------------------------------------------------------------
terminate(_Reason, _State) ->
    ok.

%%--------------------------------------------------------------------
%% @spec code_change(OldVsn, State, Extra) -> {ok, NewState}
%%--------------------------------------------------------------------
code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

%%%===================================================================
%%% Internal functions
%%%===================================================================
get_keyword_server(Keyword) ->
    ServerName = keyword_server_name(Keyword),
    ServerProcessName = list_to_atom(ServerName),
    case whereis(ServerProcessName) of
	undefined ->
            {ok, Pid} = supervisor:start_child(keyword_sup, [ServerProcessName]);
        Pid ->
            Pid
    end,
    Pid.

find_keyword_server(Keyword) ->
    ServerName = keyword_server_name(Keyword),
    ServerProcessName = list_to_atom(ServerName),
    case whereis(ServerProcessName) of
	undefined ->
	    { error, "Not found" };
        Pid ->
            { ok, Pid }
    end.    

keyword_server_name(Keyword) ->
    integer_to_list(
      erlang:phash2(
	porter:stem(
	  string:to_lower(Keyword)
	 )
       )
     ) ++ "_keyword_server".

File: lib/searchex/keyword_ser.ex
defmodule Searchex.KeywordSer do

  @moduledoc false

  use GenServer

  @doc """
  Start KeywordSer

  We start one KeywordSer for each keyword in the index.  The state looks like:

  %{"DIOCID1" => [list of positions], "DIOCID2" => [list of positions]}
  """
  def start_link(server_name \\ :server) do
    GenServer.start_link(__MODULE__, %{}, name: server_name)
  end

  def add_keyword_position(keyword, docid, position) do
    server = get_keyword_server(keyword)
    GenServer.cast(server, {:add, docid, position})
  end

  @doc """
  Perform a query

  terms is a list of strings = ["term1", "term2"]

  doc_matches looks like:
  [{"term1", %{"docid1" => [pos1, pos2], "docid2" => [pos3, pos4]}}, {"term2", %{}}]

  matches_per_term_and_doc looks like:
  %{{"term1", "docid1"} => 23, {"term1", "docid2"} => 4, ...}
  """
  def do_query(terms) when is_list(terms) do
    doc_matches = gen_doc_matches(terms)
    matches_per_term_and_doc = gen_matches_per_term_and_doc(doc_matches)
    res = Searchex.Search.Query.Bm25.doc_scores(terms, doc_matches, matches_per_term_and_doc)
    DIO.inspect :RESULTS
    DIO.inspect res
  end

  def gen_doc_matches(terms) do
    Enum.map(terms, fn(term) -> get_ids(term) end)
  end

  def gen_matches_per_term_and_doc(doc_matches) do
    Enum.reduce doc_matches, %{}, fn({term, docids}, acc1) ->
      tmp = Enum.reduce docids, %{}, fn({docid, pos_list}, acc2) ->
        Map.merge(acc2, %{{term, docid} => Enum.count(pos_list)})
      end
      Map.merge(acc1, tmp)
    end
  end

  @doc """
  Gets the list of ID's for a term.

  The returned data structure look like:
  %{"DIOCID1" => [list of positions], "DIOCID2" => [list of positions]}
  """
  def get_ids(term) do
    case find_keyword_server(term) do
      { :ok, server } -> {term, GenServer.call(server, :get_ids)};
      { :error, _ }   -> {term, %{} }
    end
  end

  @doc """
  Sets the state for a KeywordSer.  This is useful when building an in-memory
  index from a persisted index saved to disk.

  The state looks like:
  %{"DIOCID1" => [list of positions], "DIOCID2" => [list of positions]}
  """
  def set_state(server, new_state) do
    GenServer.cast(server, {:set_state, new_state})
  end

  # --- CALLBACKS

  def init(_opts) do
    {:ok, %{}}
  end

  def handle_call(:get_ids, _from, state) do
    result = state
    {:reply, result, state}
  end

  def handle_cast({:add, docid, position}, state) do
    old_list  = state[docid] || []
    new_list  = old_list ++ [position]
    new_state = Map.merge(state, %{docid => new_list})
    {:noreply, new_state}
  end

  def handle_cast({:set_state, new_state}, _state) do
    {:noreply, new_state}
  end

  def get_keyword_server(keyword) do
    name = cond do
      is_atom(keyword) -> keyword
      true             -> keyword_server_name(keyword)
    end
    Process.whereis(name) || Searchex.KeywordSup.add_child_and_return_pid(name)
  end

  def find_keyword_server(keyword) do
    name = keyword_server_name(keyword)
    case Process.whereis(name) do
      nil -> {:error, "Not found"}
      pid -> {:ok, pid}
    end
  end

  def keyword_server_name(keyword) do
    "kw_" <> keyword |> String.to_atom
  end
end
